### DISCUSSION TITLE
FeatherPad without PolKit?

### DISCUSSION THREAD

**Discussion by probonopd** [tsujan/FeatherPad#605] [Q&A]
Can FeatherPad be built without the dependency on PolKit?

How is it working on platforms where PolKit does not exist, e.g., macOS?

**Comment by tsujan** [comment]
FeatherPad doesn't depend on PolKit.

> How is it working on platforms where PolKit does not exist, e.g., macOS?

If a user tries to save a file without write permission on such platforms, the warning-bar will tell him/her that he needs Polkit.

On platforms that have a graphical front-end for su/sudo (like lxqt-sudo), FeatherPad can be run as root and save to root.

**Reply by probonopd** [reply]
Yes, I would like this PolKit message disabled and a plain message saying "This file is not writable".

**Reply by probonopd** [reply]
My non-English speaking, non-technical end users will be frightened when they see this

![image](https://user-images.githubusercontent.com/2480569/142090607-5ef7a141-d7d1-4fb8-8ca0-ab0be1aec0cb.png)

**Comment by tsujan** [comment]
> I would like this PolKit message disabled

Lack of PolKit is an exception in the Linux world (the main target of FeatherPad) and I'm sure that seeing the message on a bar, that automatically disappears after 15 seconds, doesn't hurt anyone in this special case.

> My non-English speaking, non-technical end users will be frightened when they see this

The sooner they know that something is wrong with their PolKit installation, the better. FeatherPad just shows the standard error.

**Reply by probonopd** [reply]
I am not using FeatherPad on Linux and I do not want to use PolKit (the stretch goal is to use nothing coming from the glib universe). So I take it that those messages cannot be disabled and FeatherPad will always complain about PolKit not working, even if it is not there at all? (How is the situation when FeatherPad is running on the Mac or on Windows?)

**Reply by tsujan** [reply]
> I am not using FeatherPad on Linux and I do not want to use PolKit

Then you'll lose nothing other than saving without write permission.

> and FeatherPad will always complain about PolKit not working, even if it is not there at all?

Always? No! As I said before, only when you try to save without write permission.

> How is the situation when FeatherPad is running on the Mac or on Windows?

No difference. As for macOS, I think I saw somewhere that PolKit could be installed on it but that's totally unrelated to this discussion.

**Reply by probonopd** [reply]
> only when you try to save without write permission

In this case I would like a normal error message to be shown: "You cannot save to this location because you do not have the required permissions."

No technical talk about PolKit which end users won't understand and won't be able to install anyways.

Couldn't FeatherPad do a simple check (e.g., via D-Bus or otherwise) whether PolKit is available on the system and if it is not, never mention it again?

**Reply by tsujan** [reply]
I think there's a misunderstanding here. I explain it again, by using bold, capitalized texts. Only in case the user does **NOT** have write permission but tries to save:

If Polkit is **NOT** installed, this message is shown:
```
Cannot be saved!
"pkexec" is not found. Please install Polkit!
```
If Polkit **IS** installed but there's a problem — usually due to the lack of a Polkit authentication agent — then this message is shown:
```
Cannot be saved!
[PolKit error; whatever it is]
```
The standard error is issued by Polkit. FeatherPad isn't responsible for its content.

See https://github.com/lxqt/lxqt-policykit#readme for a short explanation of Polkit authentication agent.

> Couldn't FeatherPad do a simple check (e.g., via D-Bus or otherwise) whether PolKit is available on the system and if it is not, never mention it again?

FeatherPad **DOES** check whether Polkit is available. It **SHOULD** ask the user to install it in the above-mentioned scenario.

**Reply by probonopd** [reply]
> FeatherPad DOES check whether Polkit is available. It SHOULD ask the user to install it in the above-mentioned scenario.

That is exactly what I want to avoid - asking the user to install it. 

> FeatherPad DOES check whether Polkit is available.

Good. Then it would be possible to have a run-time settings option that would disable asking the user to install PolKit, and just show normal error messages instead?

**Reply by tsujan** [reply]
> Then it would be possible to have a compile-time option

LOL! You might prefer to avoid GLib at all costs (and with it, libfm-qt, pcmanfm-qt,..., even KDE) but, I'm afraid, not tolerating a mere name that reminds you of GLib is something very very personal.

**Comment by tsujan** [comment]
BTW, the Qt widget style you've used in your screenshot has a problem. It overrides the foreground color of the widget. The warning-bar has a palette that correctly sets both background and foreground colors.

If the style belongs to Qt, you might want to report it at Qt bug tracker; otherwise, its developer(s) might need to know about it.

**Reply by probonopd** [reply]
Thanks for the hint, indeed the theme is pretty much work-in-progress.
